File: packages/gollm/common/prompts/chart_annotation.py
from entities import ChartAnnotationType

CHART_ANNOTATION_PROMPT_BASE = """
You are an agent who is an expert in Vega-Lite chart specs. Provide a Vega-Lite layer JSON object for the annotation that can be added to an existing chart spec to satisfy the provided user request.

- The Vega-Lite schema version you must use is https://vega.github.io/schema/vega-lite/v5.json.
- Assume that you don’t know the exact data points from the data.
- You must give me the single layer object that renders all the necessary drawing objects, including multiple layers within the top layer object if needed.
- When adding a label, also draw a constant line perpendicular to the axis to which you are adding the label.

Here are some examples of user requests and the corresponding JSON objects that you should provide for a chart spec:
--- EXAMPLES START ---
{examples}
--- EXAMPLES END ---

{preamble}

Give me the layer object described by the json schema found at https://vega.github.io/schema/vega-lite/v5.json to be added to the existing chart spec based on the following user instruction.
Do not respond in full sentences; only create a JSON object that satisfies the JSON schema specified in https://vega.github.io/schema/vega-lite/v5.json. Make sure you follow the JSON standard for the data types and format.

--- INSTRUCTION START ---

{instruction}

--- INSTRUCTION END ---
"""

FORECAST_CHART_ANNOTATION_EXAMPLES = """
Assuming you are adding the annotations to the following chart spec,
---- Example Chart Spec Start -----
{{
    "data": {{ "url": "data/samples.csv" }},
    "transform": [
        {{ "fold": ["price", "cost", "tax", "profit"], "as": ["variableField", "valueField"] }}
    ],
    "layer": [
        {{
            "mark": "line",
            "encoding": {{
                "x": {{ "field": "date", "type": "quantitative", "axis": {{ "title": "Day" }} }},
                "y": {{ "field": "valueField", "type": "quantitative", "axis": {{ "title": "Dollars" }} }}
            }}
        }}
    ]
}}
---- Example Chart Spec End -----
Here are some example requests and the answers:

Request:
At day 200, add a label 'important'
Answer:
{{
  "description": "At day 200, add a label 'important'",
  "layer": [
    {{
      "mark": {{
        "type": "rule",
        "strokeDash": [4, 4]
      }},
      "encoding": {{
        "x": {{ "datum": 200, "axis": {{ "title": ""}} }}
      }}
    }},
    {{
      "mark": {{
        "type": "text",
        "align": "left",
        "dx": 5,
        "dy": -5
      }},
      "encoding": {{
        "x": {{ "datum": 200, "axis": {{ "title": ""}} }}
        "text": {{"value": "important"}}
      }}
    }}
  ]
}}

Request:
Add a label 'expensive' at price 20
Answer:
{{
  "description": "Add a label 'expensive' at price 20",
  "layer": [
    {{
      "mark": {{
        "type": "rule",
        "strokeDash": [4, 4]
      }},
      "encoding": {{
        "y": {{ "datum": 20, "axis": {{ "title": ""}} }}
      }}
    }},
    {{
      "mark": {{
        "type": "text",
        "align": "left",
        "dx": 5,
        "dy": -5
      }},
      "encoding": {{
        "y": {{ "datum": 20, "axis": {{ "title": ""}} }},
        "text": {{"value": "expensive"}}
      }}
    }}
  ]
}}

Request:
Add a vertical line for the day where the price exceeds 100.
Answer:
{{
  "description": "Add a vertical line for the day where the price exceeds 100.",
  "transform": [
    {{"filter": "datum.valueField > 100"}},
    {{"aggregate": [{{"op": "min", "field": "date", "as": "min_date"}}]}}
  ],
  "layer": [
    {{
      "mark": {{
        "type": "rule",
        "strokeDash": [4, 4]
      }},
      "encoding": {{
        "x": {{"field": "min_date", "type": "quantitative", "axis": {{ "title": ""}}}}
      }}
    }},
    {{
      "mark": {{
        "type": "text",
        "align": "left",
        "dx": 5,
        "dy": -10
      }},
      "encoding": {{
        "x": {{"field": "min_date", "type": "quantitative", "axis": {{ "title": ""}}}},
        "text": {{"value": "Price > 100"}}
      }}
    }}
  ]
}}

Request:
Add a vertical line at the day where the price reaches its peak value.
Answer:
{{
  "description": "Add a vertical line at the day where the price reaches its peak value.",
  "transform": [
    {{"filter": "datum.variableField == 'price'"}},
    {{
      "joinaggregate": [{{
      "op": "max",
      "field": "valueField",
      "as": "max_price"
      }}]
    }},
    {{"filter": "datum.valueField >= datum.max_price"}}
  ],
  "layer": [
    {{
      "mark": {{
        "type": "rule",
        "strokeDash": [4, 4]
      }},
      "encoding": {{
        "x": {{"field": "date", "type": "quantitative", "axis": {{ "title": ""}}}}
      }}
    }},
    {{
      "mark": {{
        "type": "text",
        "align": "left",
        "dx": 5,
        "dy": -10
      }},
      "encoding": {{
        "x": {{"field": "date", "type": "quantitative", "axis": {{ "title": ""}}}},
        "text": {{"value": "Max Price"}}
      }}
    }}
  ]
}}
"""

QUANTILE_FORECAST_CHART_ANNOTATION_EXAMPLES="""
Assuming you are adding the annotations to the following chart spec for quantile confidence intervals,
---- Example Chart Spec Start -----
{{
    "data": {{ "url": "data/samples.csv" }},
    "layer": [
        {{
            "mark": "errorband",
            "encoding": {{
                "x": {{ "field": "x", "type": "quantitative", "axis": {{ "title": "Day" }} }},
                "y": {{ "field": "upper", "type": "quantitative", "axis": {{ "title": "Dollars" }} }}
                "y2": {{ "field": "lower", "type": "quantitative", "axis": {{ "title": "Dollars" }} }}
                "color": {{ "field": "variable", "type": "nominal" }}
                "opacity": {{ "field": "quantile", "type": "quantitative" }}
            }}
        }}
    ]
}}
---- Example Chart Spec End -----
And here are some information about the data for the example chart spec:
- data is records with following columns: x, upper, lower, variable, quantile
- When not instructed otherwise, assume that quantile is 0.5 (median) and upper and lower are the same value.
- Possible values for quantile are from 0.5 to 0.99
- Possible values for variable are price, cost, tax, profit

Here are some example requests and the answers:

Request:
At day 200, add a label 'important'
Answer:
{{
  "description": "At day 200, add a label 'important'",
  "layer": [
    {{
      "mark": {{
        "type": "rule",
        "strokeDash": [4, 4]
      }},
      "encoding": {{
        "x": {{ "datum": 200, "axis": {{ "title": ""}} }}
      }}
    }},
    {{
      "mark": {{
        "type": "text",
        "align": "left",
        "dx": 5,
        "dy": -5
      }},
      "encoding": {{
        "x": {{ "datum": 200, "axis": {{ "title": ""}} }}
        "text": {{"value": "important"}}
      }}
    }}
  ]
}}

Request:
Add a label 'expensive' at price 20
Answer:
{{
  "description": "Add a label 'expensive' at price 20",
  "layer": [
    {{
      "mark": {{
        "type": "rule",
        "strokeDash": [4, 4]
      }},
      "encoding": {{
        "y": {{ "datum": 20, "axis": {{ "title": ""}} }}
      }}
    }},
    {{
      "mark": {{
        "type": "text",
        "align": "left",
        "dx": 5,
        "dy": -5
      }},
      "encoding": {{
        "y": {{ "datum": 20, "axis": {{ "title": ""}} }},
        "text": {{"value": "expensive"}}
      }}
    }}
  ]
}}
"""

def build_prompt(chartType: ChartAnnotationType, preamble: str, instruction: str) -> str:
    examples = {
        ChartAnnotationType.FORECAST_CHART: FORECAST_CHART_ANNOTATION_EXAMPLES,
        ChartAnnotationType.QUANTILE_FORECAST_CHART: QUANTILE_FORECAST_CHART_ANNOTATION_EXAMPLES
    }[chartType]
    prompt = CHART_ANNOTATION_PROMPT_BASE.format(examples=examples, preamble=preamble, instruction=instruction)
    print(prompt)
    return prompt

File: packages/client/hmi-client/src/services/chart-annotation.ts
import _ from 'lodash';
import API from '@/api/api';
import { v4 as uuidv4 } from 'uuid';
import { b64DecodeUnicode } from '@/utils/binary';
import { ChartAnnotation, ChartAnnotationType } from '@/types/Types';
import { ForecastChartOptions } from '@/services/charts';
import { ChartSetting } from '@/types/common';

export interface LLMGeneratedChartAnnotation {
	request: string;
	layerSpec: any;
}

interface PromptPreambleInput {
	variables: string[];
	timeField: string;
	translateMap: Record<string, string>;
	axisTitle: { x: string; y: string };
}

const buildPromptPreamble = ({ variables, timeField, axisTitle, translateMap }: PromptPreambleInput) => ({
	[ChartAnnotationType.ForecastChart]: `
	Here is the information of the existing target chart spec where you need to add the annotations:
	- The existing chart follows a similar pattern as the above Example Chart Spec like:
			{
				...
				"transform": [
					{
						"fold": ${JSON.stringify(variables)},
						"as": ["variableField', "valueField"]
					}
				],
				"layer": [
					{
						...
						"encoding": {
							"x": {"field": "${timeField}", "type": "quantitative", "axis": {"title": "${axisTitle.x}"}},
							"y": {"field": "valueField", "type": "quantitative", "axis": {"title": "${axisTitle.y}"}}
						}
					}
					...
				]
			}
	- Assume all unknown variables except the time field are for the y-axis and are renamed to the valueField.
	- Make sure possible values for 'valueField' are ${JSON.stringify(variables)} and try best to translate the variables mentioned from the request to the variables for the 'valueField'.
	- Leverage this variable to human readable name mapping: ${JSON.stringify(translateMap)} if needed.
	`,

	[ChartAnnotationType.QuantileForecastChart]: `
	Here is the information of the existing target chart spec where you need to add the annotations:
		{
			...
			"layer": [
				{
					"mark": "errorband",
					"encoding": {
							"x": { "field": "x", "type": "quantitative", "axis": { "title": "${axisTitle.x}" } },
							"y": { "field": "upper", "type": "quantitative", "axis": { "title": "${axisTitle.y}" } },
							"y2": { "field": "lower", "type": "quantitative", "axis": { "title": "${axisTitle.y}" } },
							"color": { "field": "variable", "type": "nominal" },
							"opacity": { "field": "quantile", "type": "quantitative" }
					}
				}
			]
		}
	And here are some information that needs to be considered:
	- data is records with following columns: x, upper, lower, variable, quantile (x is the time field)
	- When not instructed otherwise, assume that quantile is 0.5 (median) and upper and lower are the same value.
	- Possible values for quantile are from 0.5 to 0.99
	- Possible values for variable are ${JSON.stringify(variables)}
	- Leverage this variable to human readable name mapping: ${JSON.stringify(translateMap)} for translating the variables mentioned from the request to the variables for the 'variable' field.
	`
});

export async function saveAnnotation(
	annotationLayerSpec: LLMGeneratedChartAnnotation,
	nodeId: string,
	chartId: string,
	chartType: ChartAnnotationType = ChartAnnotationType.ForecastChart
) {
	const annotation: ChartAnnotation = {
		id: uuidv4(),
		description: annotationLayerSpec.request,
		chartType,
		nodeId,
		outputId: '',
		chartId,
		layerSpec: annotationLayerSpec.layerSpec,
		llmGenerated: true,
		metadata: {}
	};
	const { data } = await API.post('/chart-annotations', annotation);
	return data as ChartAnnotation;
}

export async function deleteAnnotation(annotationId: string) {
	const { data } = await API.delete(`/chart-annotations/${annotationId}`);
	return data;
}

export async function fetchAnnotations(nodeId: string) {
	const { data } = await API.post(`/chart-annotations/search`, { nodeId });
	return data as ChartAnnotation[];
}

export async function generateChartAnnotation(
	request: string,
	chartType: ChartAnnotationType,
	timeField: string,
	variables: string[],
	options: Partial<ForecastChartOptions>
): Promise<LLMGeneratedChartAnnotation> {
	const axisTitle = {
		x: options.xAxisTitle ?? '',
		y: options.yAxisTitle ?? ''
	};
	const translateMap = _.pick(options.translationMap ?? {}, variables);
	const preamble = buildPromptPreamble({ variables, timeField, axisTitle, translateMap })[chartType];
	const { data } = await API.post(
		`/gollm/chart-annotation?mode=SYNC`,
		{ preamble, instruction: request, chartType },
		{
			headers: {
				'Content-Type': 'application/json'
			}
		}
	);
	const str = b64DecodeUnicode(data.output);
	const result = JSON.parse(str);
	const layerSpec = result.response ?? null;
	return {
		request,
		layerSpec
	};
}

/**
 * Get the chart annotation type based on the chart setting.
 * @param setting - The chart setting.
 * @returns - The chart annotation type.
 */
export function getChartAnnotationType(setting: ChartSetting) {
	if (setting.showQuantiles) {
		return ChartAnnotationType.QuantileForecastChart;
	}
	return ChartAnnotationType.ForecastChart;
}
